### PR TITLE
Fix ISR page re-rendering after revalidate expiry

### LINKED ISSUE #24806: Incremental Static Regeneration (ISR) pages not re-rendering after revalidate expiry
### What version of Next.js are you using?

10.0.5

### What version of Node.js are you using?

12.20.0

### What browser are you using?

Chrome & cURL

### What operating system are you using?

macOS

### How are you deploying your application?

next start on Azure App Service

### Describe the Bug

Under certain circumstances the Incremental Static Regeneration (ISR) logic is incorrectly prolonging the life of rendered page content beyond the revalidation expiry, potentially indefinitely.

Those circumstances:
- The site has sufficient pages (and page size) to exhaust the 50MB `LRUCache` size.
- The site has sufficient traffic that a given page may well be removed from the LRUCache within the `revalidate` period.

For context, we're experiencing this on a site with:
- Around 2,000 pages, total.
- A mean average `LRUCache` item size (page size) of 420KB.
- So around enough space for 120 pages (6%) to fit in the `LRUCache` at one time.
- Around 75k* daily page views.
- A `revalidate` duration of `300` (5 minutes).

*This is the figure from analytics. We have quite a few pages listing large quantities of internal links at a time, which with `Next/Link`'s `prefetch` behaviour amplifies this figure somewhat as far at the ISR cache is concerned.

Next's incremental cache logic makes use of an in-memory `LRUCache` (least recently used cache) with a max size of 50MB, with a fallback to the filesystem for anything not found in the `LRUCache`.

I believe there's a bug in the incremental cache logic where if a page is not present in the `LRUCache`, then when falling back to the filesystem cache the (potentially) stale file content is added back to the `LRUCache` for a full fresh revalidation period, which is a process that can repeat itself if the page keeps being pushed out of the `LRUCache` before the next request comes in after the revalidation period expires.

This creates the potential for stale content to be served in perpetuity for certain pages.

### Expected Behavior

Under the stale-while-revalidate (SWR) caching model for the pages which make use of ISR, any request for a page which was last generated more than the defined `revalidate` seconds ago should trigger the page to be regenerated in the background.

### To Reproduce

Due to its nature, this is a tricky one to reproduce. You need a site with enough ISR pages (and a big enough page size) to exhaust the 50MB `LRUCache` space, and ideally some visibility into the cache state so that you can follow whether the `LRUCache` is being hit, whether it's falling back to the filesystem, and when the cache is determining that a page is considered stale (should be regenerated by `next-server`).

To that end [I've created a branch which reduces the cache size to 1MB to make this easier to reproduce, and adds some logging to make this easier to follow](https://github.com/kyan/next.js/tree/investigate-isr-pages-not-re-rendering-after-revalidate-expiry), but it is possible to reproduce this without that branch, just harder. Setting the `revalidate` to a shorter period also makes life easier here. I found `60` (1 min) a good balance, but I believe this bug is present for any `revalidate` duration.

1. Ensure there are no cached pages - `rm -rf .next/`
2. Rebuild the site - `yarn build`
3. Serve the site - `yarn start`

At this point the `LRUCache` is empty, and the filesystem cache is populated for any pages that are prebuilt during the `yarn build`.

4. Fill up the `LRUCache`, by requesting enough pages to exhaust the 50MB (Or 1MB if you're using the [`investigate-isr-pages-not-re-rendering-after-revalidate-expiry`](https://github.com/kyan/next.js/tree/investigate-isr-pages-not-re-rendering-after-revalidate-expiry) branch) capacity.
5. Make a note of the current time. Select a test page (e.g. `/posts/foo-post`), and request it to bring it to the front of the `LRUCache`. 
6. Request enough other pages to push that page out of the cache, making sure not to request the test page you selected in step 5.

`Next/Link`'s prefetch behaviour can make life tricky here if you're using a browser, as it might sneakily request the test page if it's included as a link on one of the other pages. `curl` is likely a safer way to go here.

(See the `cacheKeys` log key for visibility of which pages are in the `LRUCache` if using the [`investigate-isr-pages-not-re-rendering-after-revalidate-expiry`](https://github.com/kyan/next.js/tree/investigate-isr-pages-not-re-rendering-after-revalidate-expiry) branch logging).

7. Wait for the `revalidate` window to expire for the test page request (step 5), if it hasn't already.
8. Make a note of the last modified timestamp from the test page's `.html` and `.json` files within the filesystem cache directory, for example:

```
$ ls -lt .next/server/pages/posts/foo-post.*
-rw-r--r--  1 user  staff   79293  4 May 13:25 .next/server/pages/posts/foo-post.json
-rw-r--r--  1 user  staff  308864  4 May 13:25 .next/server/pages/posts/foo-post.html
```

9. Request the test page.
10. Review the last modified timestamps for the test page's `.html` and `.json` files again, e.g.

```
$ ls -lt .next/server/pages/posts/foo-post.*
-rw-r--r--  1 user  staff   79293  4 May 13:25 .next/server/pages/posts/foo-post.json
-rw-r--r--  1 user  staff  308864  4 May 13:25 .next/server/pages/posts/foo-post.html
```

Note that they have not changed. The page has not been regenerated, despite the revalidation window having expired.

11. (optional) Repeat steps 6-10 again to demonstrate that this bug can potentially result in an infinite revalidation window.
12. Request the test page to bring it back to the front of the `LRUCache` again.
13. This time, do not request other pages in order to push the test page out of the `LRUCache`.
14. Wait for the `revalidate` window to expire for the most recent test page request (if it hasn't already).
15. Make a note of the last modified timestamp from the test page's `.html` and `.json` files within the Next filesystem again, e.g.

```
$ ls -lt .next/server/pages/posts/foo-post.*
-rw-r--r--  1 user  staff   79293  4 May 13:25 .next/server/pages/posts/foo-post.json
-rw-r--r--  1 user  staff  308864  4 May 13:25 .next/server/pages/posts/foo-post.html
```

16. Request the test page again.
17. Review the last modified timestamps for the test page's `.html` and `.json` files again.

```
$ ls -lt .next/server/pages/posts/foo-post.*
-rw-r--r--  1 user  staff   79293  4 May 13:31 .next/server/pages/posts/foo-post.json
-rw-r--r--  1 user  staff  308864  4 May 13:31 .next/server/pages/posts/foo-post.html
```

Note that this time the last modified timestamps have updated. The page was regenerated because it was still in the `LRUCache` when the request came in following the expiry of the revalidation window.

### PR DESCRIPTION
**Pull request by tommarshall**
* As outlined in #24806 there's currently a bug with the incremental
  static regeneration (ISR) logic incorrectly prolonging the life of
  rendered page content when the `LRUCache` is refreshed with
  potentially stale content from the filesystem cache.
* This is because if a page is not present in the `LRUCache` but is
  present in the filesystem cache, then it's re-added to the `LRUCache`
  with a fresh cache expiry calculated from the current time using the
  `revalidate` window.
* If that page continues to be pushed out of the `LRUCache` before the
  next request comes in after the revalidation window expires then this
  creates an endless loop of the same stale content from the filesystem
  being served, without it being regenerated.
* This fix addresses that issue by adding a `fromTime` argument to the
  revalidation period calculation, so that when we're falling back to
  the filesystem we can use the filesystem's `lastModified` timestamp
  for that file to anchor the revalidation calculation - rather than
  restarting from the current time in every case.
* That means that if page content from the filesystem is stale it will
  be marked as so when returned from the incremental cache `get` method
  (`isStale: true`), which tells next server to regenerate the page in
  the background.
* If however the the page content from the filesystem is still within
  the revalidate window (i.e. is not yet stale) then it will be added
  back to the LRUCache with the correct remaining duration for the
  `revalidateAfter`.

Fixes #24806

## Bug

- [x] Related issues linked using `fixes #number`
- [ ] Integration tests added

### CODE REVIEW

**Comment by tommarshall** [comment]
I appreciate this needs an integration test, but due to the nature of the issue I don't anticipate it's going to be an easy one to write.

If anyone can point me at a test that's covering similar territory (revalidation cache invalidation logic) for reference, that'd be a huge help 🙏

File: packages/next/next-server/server/incremental-cache.ts [hunk 1]
```diff
@@ -85,13 +85,15 @@ export class IncrementalCache {
...
+    pathname: string,
+    fromTime: number
+  ): number | false {
     pathname = toRoute(pathname)
 
     // in development we don't have a prerender-manifest
     // and default to always revalidating to allow easier debugging
-    const curTime = new Date().getTime()
-    if (this.incrementalOptions.dev) return curTime - 1000
+    if (this.incrementalOptions.dev) return new Date().getTime() - 1000
```
Comment (tommarshall) [inline packages/next/next-server/server/incremental-cache.ts:96]: Retain the existing behaviour for development where the current time minus 1000 is used to ensure we always regenerate pages.

File: packages/next/next-server/server/incremental-cache.ts [hunk 2]
```diff
@@ -125,18 +127,17 @@ export class IncrementalCache {
...
+        const { mtime } = await promises.stat(htmlPath)
         const pageData = JSON.parse(
           await promises.readFile(this.getSeedPath(pathname, 'json'), 'utf8')
         )
 
         data = {
           html,
           pageData,
-          revalidateAfter: this.calculateRevalidate(pathname),
+          revalidateAfter: this.calculateRevalidate(pathname, mtime.getTime()),
```
Comment (tommarshall) [inline packages/next/next-server/server/incremental-cache.ts:140]: Use the `lastModified` file timestamp from the filesystem cache, when refreshing the `LRUCache` content in `IncrementalCache.get`, as this timestamp is when the content was last generated and then written by `IncrementalCache.set` - so is the anchor point from which we should be calculating the `revalidateAfter` timestamp from here.

File: packages/next/next-server/server/incremental-cache.ts [hunk 3]
```diff
@@ -189,7 +190,7 @@ export class IncrementalCache {
     pathname = normalizePagePath(pathname)
     this.cache.set(pathname, {
       ...data,
-      revalidateAfter: this.calculateRevalidate(pathname),
+      revalidateAfter: this.calculateRevalidate(pathname, new Date().getTime()),
```
Comment (tommarshall) [inline packages/next/next-server/server/incremental-cache.ts:193]: Continue to use the current time when calculating the revalidateAfter timestamp within `IncrementalCache.set` as this should always be considered freshly generated content.

**Comment by Addono** [comment]
> If anyone can point me at a test that's covering similar territory (revalidation cache invalidation logic) for reference, and/or a suggestion for a sensible way to saturate the LRUCache that'd be a huge help 🙏

Digging a bit through the test, [`test/integration/prerender/test/index.test.js`](https://github.com/vercel/next.js/blob/canary/test/integration/prerender/test/index.test.js#L1025) seems to be the home of other SSR / IRS related tests, which spins up its own Next server serving pages from `test/integration/prerender/pages/`.

What about adding a new ISR enabled page there which allows generating large pages, e.g. have a prop with as value `"".padStart(20 * 1024 * 1024, "a")` to generate a 20Mb string. That endpoint we can use to exhaust the `LRUCache` later.

Don't think those tests really test the invalidation logic itself, what about: 
 1. Add an endpoint with 1 second ISR revalidation and as one of the props the current time (`{ revalidate: 1, props: { time: Date.now().toString() } } `).
 2. Then retrieve this page at the start of the test and extract the time from the JSON encoded in the body. Similar as to: https://github.com/vercel/next.js/blob/f18ce55d8ac85f8abaa51bcd8a663ef5c8e78d28/test/integration/prerender/test/index.test.js#L428-L431.
 3. Access endpoint with our large pages (described above) a couple of times.
 4. Wait 1 second, like: https://github.com/vercel/next.js/blob/f18ce55d8ac85f8abaa51bcd8a663ef5c8e78d28/test/integration/prerender/test/index.test.js#L1025
 5. Access our endpoint from (1) again and check if the time prop has increased by at least 1000ms.

---

Just jotting down the structure of the test I have in mind as code, something like this:

```tsx
// Get orignal page
const startHtml = await renderViaHTTP(appPort, '/blog/isr-endpoint')

// Exhaust LRU cache
await renderViaHTTP(appPort, '/blog/large-endpoint/1')
await renderViaHTTP(appPort, '/blog/large-endpoint/2') 

// Wait until the ISR generated page needs to be revalidated
await waitFor(1000) 

// Validate that our ISR page was regenerated
const newHtml = await renderViaHTTP(appPort, '/blog/isr-endpoint')

const $start = cheerio.load(startHtml)
const $new = cheerio.load(newHtml)

const startTime = parseInt(JSON.parse($('#__NEXT_DATA__').text()).props.pageProps.time)
const newTime = parseInt(JSON.parse($('#__NEXT_DATA__').text()).props.pageProps.time)

// The new page should have been generated more than 1s later than
// the original page we retrieved.
expect(newTime).toBeGreaterThanOrEqual(startTime + 1000)
```

**Review by ijjk — COMMENTED** [review]
Hi, we can add a test for this by adding a private `process.env.__NEXT_TEST_MAX_ISR_CACHE` variable to [this line](https://github.com/vercel/next.js/blob/46a85b4e6f48c35d785a113434f038b7879313be/packages/next/next-server/server/incremental-cache.ts#L75) which sets the cache to a low size so it's easier to max out, this can be in it's own test suite separate of `test/integration/prerender` so it's easier to special case.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ba5b251f2e23d972bbe40579a5be3622f4c51397

**test/integration/prerender/test/index.test.js** 
- SSG Prerender > production mode > should handle revalidating HTML correctly
- SSG Prerender > serverless mode > should handle revalidating HTML correctly

<details>
<summary>Expand output</summary>

  ● SSG Prerender › serverless mode › should handle revalidating HTML correctly

    expect(received).toBe(expected) // Object.is equality

    Expected: "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><noscript data-n-css=\"\"></noscript><link rel=\"preload\" href=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" as=\"script\"/></head><body><div id=\"__next\"><p>Post: post-2</p><p>Comment: comment-2</p><span>time: 1624980216779</span><a id=\"home\" href=\"/\">to home</a></div><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{\"post\":\"post-2\",\"comment\":\"comment-2\",\"time\":1624980216779},\"__N_SSG\":true},\"page\":\"/blog/[post]/[comment]\",\"query\":{\"post\":\"post-2\",\"comment\":\"comment-2\"},\"buildId\":\"pBV3Ragrs9BTS5Xyws3W9\",\"runtimeConfig\":{},\"isFallback\":false,\"gsp\":true,\"scriptLoader\":[]}</script><script nomodule=\"\" src=\"/_next/static/chunks/polyfills-6d4f5fa3f9e199418011.js\"></script><script src=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" async=\"\"></script><script src=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" async=\"\"></script><script src=\"/_next/static/pBV3Ragrs9BTS5Xyws3W9/_buildManifest.js\" async=\"\"></script><script src=\"/_next/static/pBV3Ragrs9BTS5Xyws3W9/_ssgManifest.js\" async=\"\"></script></body></html>"
    Received: "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><noscript data-n-css=\"\"></noscript><script defer=\"\" nomodule=\"\" src=\"/_next/static/chunks/polyfills-6d4f5fa3f9e199418011.js\"></script><script src=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" defer=\"\"></script><script src=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" defer=\"\"></script><script src=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" defer=\"\"></script><script src=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" defer=\"\"></script><script src=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" defer=\"\"></script><script src=\"/_next/static/pBV3Ragrs9BTS5Xyws3W9/_buildManifest.js\" defer=\"\"></script><script src=\"/_next/static/pBV3Ragrs9BTS5Xyws3W9/_ssgManifest.js\" defer=\"\"></script></head><body><div id=\"__next\"><p>Post: post-2</p><p>Comment: comment-2</p><span>time: 1624980246255</span><a id=\"home\" href=\"/\">to home</a></div><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{\"post\":\"post-2\",\"comment\":\"comment-2\",\"time\":1624980246255},\"__N_SSG\":true},\"page\":\"/blog/[post]/[comment]\",\"query\":{\"post\":\"post-2\",\"comment\":\"comment-2\"},\"buildId\":\"pBV3Ragrs9BTS5Xyws3W9\",\"runtimeConfig\":{},\"isFallback\":false,\"gsp\":true,\"scriptLoader\":[]}</script></body></html>"

      1646 |
      1647 |         let newHtml = await renderViaHTTP(appPort, route)
    > 1648 |         expect(newHtml).toBe(initialHtml)
           |                         ^
      1649 |
      1650 |         await waitFor(2 * 1000)
      1651 |         await renderViaHTTP(appPort, route)

      at Object.<anonymous> (integration/prerender/test/index.test.js:1648:25)
          at runMicrotasks (<anonymous>)

  ● SSG Prerender › production mode › should handle revalidating HTML correctly

    expect(received).toBe(expected) // Object.is equality

    Expected: "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><noscript data-n-css=\"\"></noscript><link rel=\"preload\" href=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" as=\"script\"/></head><body><div id=\"__next\"><p>Post: post-2</p><p>Comment: comment-2</p><span>time: 1624980379085</span><a id=\"home\" href=\"/\">to home</a></div><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{\"post\":\"post-2\",\"comment\":\"comment-2\",\"time\":1624980379085},\"__N_SSG\":true},\"page\":\"/blog/[post]/[comment]\",\"query\":{\"post\":\"post-2\",\"comment\":\"comment-2\"},\"buildId\":\"pDkgbf1UqKmRDTT2Kg4zo\",\"isFallback\":false,\"gsp\":true,\"scriptLoader\":[]}</script><script nomodule=\"\" src=\"/_next/static/chunks/polyfills-6d4f5fa3f9e199418011.js\"></script><script src=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" async=\"\"></script><script src=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" async=\"\"></script><script src=\"/_next/static/pDkgbf1UqKmRDTT2Kg4zo/_buildManifest.js\" async=\"\"></script><script src=\"/_next/static/pDkgbf1UqKmRDTT2Kg4zo/_ssgManifest.js\" async=\"\"></script></body></html>"
    Received: "<!DOCTYPE html><html><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width\"/><meta name=\"next-head-count\" content=\"2\"/><noscript data-n-css=\"\"></noscript><link rel=\"preload\" href=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" as=\"script\"/><link rel=\"preload\" href=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" as=\"script\"/></head><body><div id=\"__next\"><p>Post: post-2</p><p>Comment: comment-2</p><span>time: 1624980409058</span><a id=\"home\" href=\"/\">to home</a></div><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{\"post\":\"post-2\",\"comment\":\"comment-2\",\"time\":1624980409058},\"__N_SSG\":true},\"page\":\"/blog/[post]/[comment]\",\"query\":{\"post\":\"post-2\",\"comment\":\"comment-2\"},\"buildId\":\"pDkgbf1UqKmRDTT2Kg4zo\",\"isFallback\":false,\"gsp\":true,\"scriptLoader\":[]}</script><script nomodule=\"\" src=\"/_next/static/chunks/polyfills-6d4f5fa3f9e199418011.js\"></script><script src=\"/_next/static/chunks/webpack-43375c3d74aa814440ba.js\" async=\"\"></script><script src=\"/_next/static/chunks/framework-999d43a842a85dee1b3e.js\" async=\"\"></script><script src=\"/_next/static/chunks/main-94d212e1d8fd30d30c9e.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/_app-f69ca0a0bd1cca597909.js\" async=\"\"></script><script src=\"/_next/static/chunks/pages/blog/%5Bpost%5D/%5Bcomment%5D-ef51ff0daa48abdcae49.js\" async=\"\"></script><script src=\"/_next/static/pDkgbf1UqKmRDTT2Kg4zo/_buildManifest.js\" async=\"\"></script><script src=\"/_next/static/pDkgbf1UqKmRDTT2Kg4zo/_ssgManifest.js\" async=\"\"></script></body></html>"

      1646 |
      1647 |         let newHtml = await renderViaHTTP(appPort, route)
    > 1648 |         expect(newHtml).toBe(initialHtml)
           |                         ^
      1649 |
      1650 |         await waitFor(2 * 1000)
      1651 |         await renderViaHTTP(appPort, route)

      at Object.<anonymous> (integration/prerender/test/index.test.js:1648:25)
          at runMicrotasks (<anonymous>)

</details>

**Comment by Stefwint** [comment]
@ijjk Are you planning on adding this fix to the next release? We're also encountering this issue on a production website where the pages do not get revalidated after a while.

**Comment by ijjk** [comment]
@Stefwint we plan to get this landed as soon as the the PR checklist has been met, currently waiting on a test case being added to prevent regression. This PR also can't be updated by maintainers as it seems that was disabled for this PR.

**Pull request closed without merging**